### ISSUE TITLE
flutter build --release fails with snapshot generator error exit code -1073740791 or -6

### ISSUE THREAD

**Issue by Moncader** [flutter/flutter#35613]
I have a very large application and have been using the flutter stable channel (1.5.4-hotfix.2) for a long while now.

I am looking to upgrade to the current beta (1.7.8+hotfix.2) as it fixes several issues we've had around ios keyboard lag issues.

Upon upgrading and testing the app, everything works fine in debug mode, however all release builds for apk or ios fail with the dart snapshot generation failing due to unreachable code as shown in the log below.

The application is very large so I'm not sure we can get a minimal test case. I may try later though.

The issue occurs on both windows and mac build machines.

## Steps to Reproduce

1. `flutter build apk (or ios) --release`

## Logs

```
flutter build apk --release --verbose --flavor primarydevarm64debugsign --target-platform android-arm64 -t lib/main_dev.dart
[  +25 ms] executing: [xxxxxxxxxxxxxxxxxxxx/flutter/] git log -n 1 --pretty=format:%H
[  +30 ms] Exit code 0 from: git log -n 1 --pretty=format:%H
[        ] 2e540931f73593e35627592ca4f9a4ca3035ed31
[        ] executing: [xxxxxxxxxxxxxxxxxxxx/flutter/] git describe --match v*.*.* --first-parent --long --tags
[  +16 ms] Exit code 0 from: git describe --match v*.*.* --first-parent --long --tags
[        ] v1.7.8+hotfix.2-0-g2e540931f
[   +9 ms] executing: [xxxxxxxxxxxxxxxxxxxx/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[  +11 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/beta
[        ] executing: [xxxxxxxxxxxxxxxxxxxx/flutter/] git ls-remote --get-url origin
[  +10 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +39 ms] executing: [xxxxxxxxxxxxxxxxxxxx/flutter/] git rev-parse --abbrev-ref HEAD
[  +11 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] beta
[  +71 ms] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +3 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[ +123 ms] Initializing gradle...
[   +7 ms] Using gradle from xxxxxxxxxxxxxxxxxx/android/gradlew.
[ +483 ms] executing: xxxxxxxxxxxxxxxxxx/android/gradlew -v
[ +597 ms] 
                    ------------------------------------------------------------
                    Gradle 4.10.2
                    ------------------------------------------------------------

                    Build time:   2018-09-19 18:10:15 UTC
                    Revision:     b4d8d5d170bb4ba516e88d7fe5647e2323d791dd

                    Kotlin DSL:   1.0-rc-6
                    Kotlin:       1.2.61
                    Groovy:       2.4.15
                    Ant:          Apache Ant(TM) version 1.9.11 compiled on March 23 2018
                    JVM:          1.8.0_161 (Oracle Corporation 25.161-b12)
                    OS:           Mac OS X 10.13.6 x86_64
[   +7 ms] Initializing gradle... (completed in 1.1s)
[   +7 ms] Resolving dependencies...
[        ] executing: [xxxxxxxxxxxxxxxxxx/android/] xxxxxxxxxxxxxxxxxx/android/gradlew app:properties
[+10147 ms] Starting a Gradle Daemon (subsequent builds will be faster)

                     > Configure project :app
xxxxxxxxxxxxxxxxxxxxxxxxxx lots of gradle related tasks
xxxxxxxxxxxxxxxxxxxxxxxxxx many many more registerResGeneratingTask messages
[        ] registerResGeneratingTask is deprecated, use registerGeneratedResFolders(FileCollection)
xxxxxxxxxxxxxxxxxxxxxxxxxx many many more Transforms....
[        ] > Transform multidex.aar (androidx.multidex:multidex:2.0.1) with AarTransform
[+30004 ms] > Task :app:compileflutterBuildPrimaryDevArm64DebugsignReleaseArm64
[+5800 ms] ../../third_party/dart/runtime/vm/compiler/aot/precompiler.cc: 2371: error: unreachable code
[        ] Dart snapshot generator failed with exit code -6
[        ] Snapshotting exited with non-zero exit code: -6
[  +90 ms] > Task :app:compileflutterBuildPrimaryDevArm64DebugsignReleaseArm64 FAILED
[        ] 12 actionable tasks: 1 executed, 11 up-to-date
[        ]          *********************************************************
[        ] WARNING: This version of android_intent will break your Android build if it or its dependencies aren't compatible with AndroidX.
[        ]          See https://goo.gl/CP92wY for more information on the problem and how to fix it.
[        ]          This warning prints for all Android build failures. The real root cause of the error may be unrelated.
[        ]          *********************************************************
[  +56 ms] FAILURE: Build failed with an exception.
[        ] * Where:
[        ] Script 'xxxxxxxxxxxxx/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 638
[        ] * What went wrong:
[        ] Execution failed for task ':app:compileflutterBuildPrimaryDevArm64DebugsignReleaseArm64'.
[        ] > Process 'command 'xxxxxxxxxxxxx/flutter/bin/flutter'' finished with non-zero exit value 1
[        ] * Try:
[        ] Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
[        ] * Get more help at https://help.gradle.org
[        ] BUILD FAILED in 39s
[ +391 ms] Running Gradle task 'assemblePrimarydevarm64debugsignRelease'... (completed in 40.1s)
[        ] *******************************************************************************************
[        ] The Gradle failure may have been because of AndroidX incompatibilities in this Flutter app.
[        ] See https://goo.gl/CP92wY for more information on the problem and how to fix it.
[        ] *******************************************************************************************
[  +11 ms] "flutter apk" took 55,161ms.
[        ] "flutter apk" took 55,161ms.
Gradle task assemblePrimarydevarm64debugsignRelease failed with exit code 1

#0      throwToolExit (package:flutter_tools/src/base/common.dart:28:3)
#1      _buildGradleProjectV2 (package:flutter_tools/src/android/gradle.dart:507:5)
#2      _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:77:64)
#3      _rootRunUnary (dart:async/zone.dart:1132:38)
#4      _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#5      _FutureListener.handleValue (dart:async/future_impl.dart:126:18)
#6      Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:639:45)
#7      Future._propagateToListeners (dart:async/future_impl.dart:668:32)
#8      Future._complete (dart:async/future_impl.dart:473:7)
#9      _SyncCompleter.complete (dart:async/future_impl.dart:51:12)
#10     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:28:18)
#11     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:294:13)
#12     runCommandAndStreamOutput (package:flutter_tools/src/base/process.dart)
#13     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:77:64)
#14     _rootRunUnary (dart:async/zone.dart:1132:38)
#15     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#16     _FutureListener.handleValue (dart:async/future_impl.dart:126:18)
#17     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:639:45)
#18     Future._propagateToListeners (dart:async/future_impl.dart:668:32)
#19     Future._completeWithValue (dart:async/future_impl.dart:483:5)
#20     Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:513:7)
#21     _rootRun (dart:async/zone.dart:1124:13)
#22     _CustomZone.run (dart:async/zone.dart:1021:19)
#23     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
#24     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
#25     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#26     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#27     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:116:13)
#28     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:173:5)
```

```
flutter doctor -v
[✓] Flutter (Channel beta, v1.7.8+hotfix.2, on Mac OS X 10.13.6 17G7024, locale en-JP)
    • Flutter version 1.7.8+hotfix.2 at xxxxxxxxxxxxxxx/flutter
    • Framework revision 2e540931f7 (2 days ago), 2019-07-02 09:31:07 -0700
    • Engine revision b1cb0d9e9b
    • Dart version 2.4.0

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at xxxxxxxxxxxxxxxxx/android-sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = xxxxxxxxxxxxxxxxx/android-sdk
    • Java binary at: /Library/Java/JavaVirtualMachines/jdk1.8.0_161.jdk/Contents/Home/bin/java
    • Java version Java(TM) SE Runtime Environment (build 1.8.0_161-b12)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • CocoaPods version 1.7.2

[✓] iOS tools - develop for iOS devices
    • ios-deploy 1.9.4

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/setup/#android-setup for detailed instructions).

[✓] VS Code (version 1.35.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.2.0

[✓] Connected device (1 available)
    • xxxxxxxxxxxxxxxxxxxxxxxxxxxx

! Doctor found issues in 1 category.
```

**Comment by mraleph** [comment]
We seem to be retrying invalid speculation in a loop. We would need more information to debug this. 

First we will need to establish the name of the function which causes this. To get this you would need to run 

```console
$ flutter build aot --release --extra-gen-snapshot-options=--trace-compiler
```

Right before the crash you should see a sequence of printings mentioning function name:

```
Retrying compilation ..., suppressing inlining of deopt_id:...
```

This is the offending function. 

Then we would need IR for this function, which can be obtained by doing 

```console
$ flutter build aot --release --extra-gen-snapshot-options=--print-flow-graph-optimized,--print-flow-graph-filter=name_of_a_function
```

where `name_of_a_function` is the name of the offending function. 

/cc @mkustermann @alexmarkov @a-siva

**Comment by Moncader** [comment]
@mraleph 

Thanks for the quick response.
Here's the output of the above commands.

```
Retrying compilation package:xxxxxxxxxxxxxxxx.dart_TwoPieceEllipsisText__buildTextPainter@989491676, suppressing inlining of deopt_id:-1
Retrying compilation package:xxxxxxxxxxxxxxxx.dart_TwoPieceEllipsisText__buildTextPainter@989491676, suppressing inlining of deopt_id:-1
```

That's the function.
And the IL was large so I put it in a gist
https://gist.github.com/Moncader/b06ebe70e9fd074766cba81f7cf17c31

**Comment by mraleph** [comment]
Thanks @Moncader !

The problematic part of the IR (I think) is this one:

```
612: B134[target]:124 ParallelMove v0 <- DS-24                               
614:     v1206 <- Unbox(v1180 T{_Double}) T{_Double}                         
616:     Branch if EqualityCompare(v1206 == v1252) T{bool} goto (135, 136)   
618: B135[target]:134                                                        
620:     v538 <- Int64ToDouble(v231 T{_Double}) T{_Double}                   
622:     v525 <- BinaryDoubleOp(+, v1206, v538) T{_Double}                   
624:     v527 <- BinaryDoubleOp(*, v525, v1206) T{_Double}                   
626:     v529 <- BinaryDoubleOp(*, v527, v538) T{_Double}                    
627:     ParallelMove DS-27 <- v1                                            
628:     v1210 <- Box(v529) T{_Double}                                       
629:     ParallelMove r0 <- r5, r1 <- C, r2 <- C, S-16 <- r5                 
630:     AssertAssignable:142(v1210, int, '', instantiator_type_args(v0), function_type_args(v0)) T{_Double}
631:     ParallelMove r0 <- S-16                                             
632:     v1317 <- UnboxInt64(v1210 T{_Double})                               
634:     ParallelMove r0 <- r1 goto:146 B149   
```

(v1206 + v538) * v1206 * v538

We have a `Double` value that we are trying to unbox into `Int64` value - which is impossible and causes compilation to fail. At the same time at 630 we have a type check that would always fail - so anything dominated by it should be dead. 

I am not entirely sure how this sort of code arouse from Dart source. Do you happen to recognise this expression `(b + a) * b * a` somewhere in the method?

**Comment by Moncader** [comment]
@mraleph 

I'm able to share this function with you. So here you go.
https://gist.github.com/Moncader/17795f8203e553f10da76a84bd5860ce

**Comment by mraleph** [comment]
Thanks @Moncader - the crash reproduces easily with this function. I added some stuff to make it compilable

https://gist.github.com/mraleph/f2e6b128b34820bcc3d6aa1ea20057e9

Workaround: If you want to unblock yourself you can replace the call to `min` with your own method:

```dart
_min(int a, int b) => a <= b ? a : b;
 
final tNextFirstTextLength = _min(
        tLastTextPosition.offset - endText.length,
        tFirstText.length - 1,
      );
```

That should fix the crash. 

The underlying reason is that `AssertAssignable(BoxDouble(...), int)` creates a compile type which is an intersection of `int` and `_Double` (has abstract type of `int` but `kDoubleCid`).

This later causes a `Phi` into which it flows to be inferred as `int` phi and unboxed. 

This produces an unbox operation which tries to create a deoptimization stub, which forces reoptimization. 

I see two simple potential ways to fix it:

The most robust way is to fix `AssertAssignableInstr::ComputeType` to not create impossible types. This might have negative implications on performance, but I think it is better overall.

**Comment by Moncader** [comment]
Excellent. I'll try this out tomorrow (Japan time)!

**Comment by mraleph** [comment]
This should be fixed upstream by a85f6ff34830de7e15c6bbce34f869e699f8e1b4 waiting for roll to pick it up.

**Comment by Hixie** [comment]
@mraleph Thanks for the prompt attention to this bug. Do you have any suggestions for workarounds for people who don't want to switch to dev builds?

**Comment by Moncader** [comment]
As a memo, the workaround of a custom implementation of `min` worked fine in our case.

Also, you don't need the custom implementation.
Dart's math.min function is actually generic, and instead of `math.min(a, b)` you use `math.min<int>(a, b)` the compile error does not happen.

**Comment by mraleph** [comment]
@Hixie The problem is that it is very hard to identify offending code - it requires understanding of debugging flags and decent knowledge of Dart AOT pipeline to be able to locate the function which causes things to go off the rails. So providing universal workaround is not possible.

**Comment by mkustermann** [comment]
@Hixie / @mraleph Is there a process for cherry-picking fixes to stable?

**Comment by Hixie** [comment]
We have a policy of not hot-fixing patches that don't have tests.

How common is this? Do we have metrics showing how common it is? If not, that's probably the first thing to add, do we can quantify it.

**Comment by Hixie** [comment]
@mraleph This landed, right? Should we close this bug?

**Comment by stefanrusek** [comment]
I just lost a day of progress, due to this bug. It is still happening in Dart 2.9.0 (build 2.9.0-3.0.dev b0d35855d8)

**Comment by mraleph** [comment]
> I just lost a day of progress, due to this bug. 

Unlikely that it's this bug - it's some other bug that looks like this one, so you should file a new issue for it and provide as much information as possible. 

Also it's unclear how you can loose any progress due to this - it just crashes release builds. Your code remains intact.
